### ISSUE TITLE
getIndex(..., keep="rda") should be allowed

### ISSUE THREAD

**Issue by dankelley** [ArgoCanada/argoFloats#587]
I am working on the things we talked about yesterday.  We don't have issues for all the elements, so I won't bother trying to cross-reference.

The gist is that @richardsc was hoping for a way to (a) save downloads and (b) speed up `getIndex()`.  Using `keep=TRUE` does the former but it only speeds things up by reducing time by about 50% or so (lots of tests below, in Details).  The short story is:

**Case A: download and create RDA file**

* download .gz file: 18.2 seconds
* check out-of-range longitudes: 3.5 seconds
* decode times: 0.6 seconds
* saving RDA file: 15.1 seconds

so the download is only about 1/2 of the time.

**Case B: use existing RDA file**

* loading takes about 2 seconds.

Click the next word to see the details of the long story.  I don't understand the negative elapsed times.

<details>

```
> library(argoFloats)
> i <- getIndex("core", debug=3)
getIndex(filename="core", server=c("ifremer-https", "usgodae)", destdir="~/data/argo") { ...
  clock0= 2023-03-02 11:28:49.714 
  Converted filename='core' to filename='ar_index_global_prof.txt.gz'.
  Set destfileRda="~/data/argo/ar_index_global_prof.rda".
  getIndex() is about to check the cache
  argoFloatsIsCached(name="core") {
    returning FALSE 
  } argoFloatsIsCached()
  This destfileRda already exists, and its age is 0.018 days.
  Using existing destfileRda, since its age is under 1 days.
  About to load '~/data/argo/ar_index_global_prof.rda'...
  ... took 10.2565 seconds
  Storing this index in a memory cache, for this R session.
  argoFloatsStoreInCache(name="core")
  } argoFloatsStoreInCache()
} # getIndex() took 10.28574 seconds (returning from location A)
> i <- getIndex("core", debug=3)
getIndex(filename="core", server=c("ifremer-https", "usgodae)", destdir="~/data/argo") { ...
  clock0= 2023-03-02 11:30:05.731 
  Converted filename='core' to filename='ar_index_global_prof.txt.gz'.
  Set destfileRda="~/data/argo/ar_index_global_prof.rda".
  getIndex() is about to check the cache
  argoFloatsIsCached(name="core") {
    returning TRUE 
  } argoFloatsIsCached()
  argoFloatsWhenCached(name="core") {
    returning 2023-03-02 11:29:00.012 
  } argoFloatsWhenCached()
  cacheAge= 0.0007606851 days; age= 1 days
  cacheAge < age, so using an index that is cached in memory for this R session
  argoFloatsGetFromCache(name="core")
  } # argoFloatsGetFromCached()
  now= 2023-03-02 11:30:05.736 
} # getIndex() took -0.7317729 seconds (returning from location B)
> i <- getIndex("core", age=0, debug=3)
getIndex(filename="core", server=c("ifremer-https", "usgodae)", destdir="~/data/argo") { ...
  clock0= 2023-03-02 11:30:13.934 
  Converted filename='core' to filename='ar_index_global_prof.txt.gz'.
  Set destfileRda="~/data/argo/ar_index_global_prof.rda".
  getIndex() is about to check the cache
  argoFloatsIsCached(name="core") {
    returning TRUE 
  } argoFloatsIsCached()
  argoFloatsWhenCached(name="core") {
    returning 2023-03-02 11:29:00.012 
  } argoFloatsWhenCached()
  cacheAge= 0.0008556207 days; age= 0 days
  cacheAge >= age, so the cached index is not being used
  This destfileRda already exists, and its age is 0.019 days.
  Must update destfileRda, since its age exceeds 0.00 days
  Allocated temporary file
    '/var/folders/8b/l4h64m1j22v5pb7vj049ff140000gn/T//Rtmph6ZCL9/argo2b511f93762.gz'.
  About to try downloading an index file ...
     trying 'https://data-argo.ifremer.fr/ar_index_global_prof.txt.gz'.
Downloading 'https://data-argo.ifremer.fr/ar_index_global_prof.txt.gz'
  ... took 18.24131 seconds
  About to read the header at the start of the index file ...
  ... took 0.002747059 seconds
  Reading index file contents (can be slow).
  Setting out-of-range longitude and latitude to NA ...
  ... took 3.541197 seconds
  Decoding times ...
  ... took 0.6489911 seconds
  Saving the RDA file ...
  ... took 15.07516 seconds
  Cleaning up ...
  ... took 0.005445004 seconds
  storing newly-read index in memory for this R session ...
  argoFloatsStoreInCache(name="core")
  } argoFloatsStoreInCache()
} getIndex() took  57.90104 seconds (returning from location C)
> i <- getIndex("core", age=0, debug=3)
getIndex(filename="core", server=c("ifremer-https", "usgodae)", destdir="~/data/argo") { ...
  clock0= 2023-03-02 11:31:29.100 
  Converted filename='core' to filename='ar_index_global_prof.txt.gz'.
  Set destfileRda="~/data/argo/ar_index_global_prof.rda".
  getIndex() is about to check the cache
  argoFloatsIsCached(name="core") {
    returning TRUE 
  } argoFloatsIsCached()
  argoFloatsWhenCached(name="core") {
    returning 2023-03-02 11:31:11.835 
  } argoFloatsWhenCached()
  cacheAge= 0.0001998926 days; age= 0 days
  cacheAge >= age, so the cached index is not being used
  This destfileRda already exists, and its age is 0 days.
  Must update destfileRda, since its age exceeds 0.00 days
  Allocated temporary file
    '/var/folders/8b/l4h64m1j22v5pb7vj049ff140000gn/T//Rtmph6ZCL9/argo2b5a30c26.gz'.
  About to try downloading an index file ...
     trying 'https://data-argo.ifremer.fr/ar_index_global_prof.txt.gz'.
Downloading 'https://data-argo.ifremer.fr/ar_index_global_prof.txt.gz'
  ... took 24.59754 seconds
  About to read the header at the start of the index file ...
  ... took 0.001194 seconds
  Reading index file contents (can be slow).
  Setting out-of-range longitude and latitude to NA ...
  ... took 0.823998 seconds
  Decoding times ...
  ... took 1.146755 seconds
  Saving the RDA file ...
  ... took 15.29065 seconds
  Cleaning up ...
  ... took 0.007331848 seconds
  storing newly-read index in memory for this R session ...
  argoFloatsStoreInCache(name="core")
  } argoFloatsStoreInCache()
} getIndex() took  55.9163 seconds (returning from location C)
> i <- getIndex("core", age=1, debug=3)
getIndex(filename="core", server=c("ifremer-https", "usgodae)", destdir="~/data/argo") { ...
  clock0= 2023-03-02 11:32:41.812 
  Converted filename='core' to filename='ar_index_global_prof.txt.gz'.
  Set destfileRda="~/data/argo/ar_index_global_prof.rda".
  getIndex() is about to check the cache
  argoFloatsIsCached(name="core") {
    returning TRUE 
  } argoFloatsIsCached()
  argoFloatsWhenCached(name="core") {
    returning 2023-03-02 11:32:25.017 
  } argoFloatsWhenCached()
  cacheAge= 0.0001944309 days; age= 1 days
  cacheAge < age, so using an index that is cached in memory for this R session
  argoFloatsGetFromCache(name="core")
  } # argoFloatsGetFromCached()
  now= 2023-03-02 11:32:41.816 
} # getIndex() took -0.8128741 seconds (returning from location B)
```

**Comment by dankelley** [comment]
As the previous shows, using an existing RDA file can be quite a lot quicker than using an existing tgz file.  Therefore, the `keep` mechanism may still have `getIndex()` being too slow for some purposes.

But `getIndex()` *always* saves an RDA file.  Maybe the solution to the problem that Clark mentions is to archive/share that file.

**Comment by richardsc** [comment]
Yes, actually I had gone into this assuming that `getIndex()` _was_ able to read the information stored in the RDA file. But what I discovered was that it's stored simply as a `data.frame` rather than as an argoFloats `index` object.

So, I agree that being able to either read the RDA index file with `getIndex()` OR being able to simply load the object from the RDA would be sufficient to speed up index loading (if we don't require the latest up-to-the-minute index).

**Comment by dankelley** [comment]
Actually, the RDA holds a list but you're correct -- it is not tied to the `argoFloats` package (see 'Example' below).  This is on purpose -- the idea is that folks can do whatever they want with the RDA file.  They don't have to use or install the `argoFloats` package.  

Essentially, the local RDA helps the user by reducing the need to wait about a minute for a download, then a reading of the data, then a decoding of times, then a QC check on locations, and so forth.  For the example we talked about over Z, that change from O(minute) to O(second) might be helpful.

I will work on this stuff later this afternoon.

PS. I am getting weird things with RStudio lately.  I don't know whether it's RStudio or my beta OS.  For example, if I click that 'Build / clean and install' button, it doesn't always seem to actually update it.  I find I need to kill RS and restart it.  Strange.  I am also finding some negative elapsed times in my tests, which is even stranger.  All of this makes it a bit slow to do the work.  I am adding a lot of instrumentation to monitor timings, because another thing I want to do is to find ways to speed things up.  For example, I tested several ways to decode times in the past but I don't know which is the fastest with R as it exists today (with the new font encoding and so forth, which I think can slow down `grep()` and similar functions a lot).

Example:
```R
> system.time(load("~/data/argo/ar_index_global_prof.rda"))
   user  system elapsed 
  3.889   0.020   2.916 
> ls()
[1] "argoFloatsIndex"
> str(argoFloatsIndex, 1)
List of 4
 $ server : chr "https://data-argo.ifremer.fr"
 $ ftpRoot: chr [1:2] "ftp://ftp.ifremer.fr/ifremer/argo/dac" "ftp://usgodae.org/pub/outgoing/argo/dac"
 $ header : chr [1:8] "# Title : Profile directory file of the Argo Global Data Assembly Center" "# Description : The directory file describes all individual profile files of the argo GDAC ftp site." "# Project : ARGO" "# Format version : 2.0" ...
 $ index  :'data.frame':	2773327 obs. of  8 variables:
```
